source/lac/sparse_direct.cc

```diff
@@ -1043,42 +1152,88 @@ SparseDirectMUMPS::initialize(const Matrix &matrix)
+  // Check that the matrix has at least one nonzero element:
   Assert(nnz != 0, ExcNotInitialized());
   Assert(n == dst.size(), ExcMessage("Destination vector has the wrong size."));
   Assert(n == src.size(), ExcMessage("Source vector has the wrong size."));
 
-  // Hand over right-hand side
-  copy_rhs_to_mumps(src);
 
-  // Start solver
-  id.job = 3;
-  dmumps_c(&id);
-  copy_solution(dst);
-}
+  if constexpr (std::is_same_v<VectorType, Vector<double>>)
+    {
+      // Centralized assembly for serial vectors.
+
+      // Hand over right-hand side
+      copy_rhs_to_mumps(src);
 
+      // Start solver
+      id.job = 3;
+      dmumps_c(&id);
+      copy_solution(dst);
+    }
+  else if constexpr (std::is_same_v<VectorType,
+                                    TrilinosWrappers::MPI::Vector> ||
+                     std::is_same_v<VectorType, PETScWrappers::MPI::Vector>)
+    {
+      if constexpr (std::is_same_v<VectorType, TrilinosWrappers::MPI::Vector>)
+        id.rhs_loc = const_cast<double *>(src.begin());
+      else if constexpr (std::is_same_v<VectorType, PETScWrappers::MPI::Vector>)
+        {
+          PetscScalar *local_array;
+          VecGetArray(
+            const_cast<PETScWrappers::MPI::Vector &>(src).petsc_vector(),
+            &local_array);
+          id.rhs_loc = local_array;
+          VecRestoreArray(
+            const_cast<PETScWrappers::MPI::Vector &>(src).petsc_vector(),
+            &local_array);
+        }
+
+
+      if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+        {
+          rhs.resize(id.lrhs_loc);
+          id.rhs = rhs.data();
+        }
 
+      // Start solver
+      id.job = 3;
+      dmumps_c(&id);
 
+      // Copy solution into the given vector
+      // For MUMPS with centralized solution (icntl[20]=0), the solution is only
+      // on the root process (0) and needs to be distributed to all processes
+      if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+        {
+          // Set all the values in the dst vector
+          for (size_type i = 0; i < n; ++i)
+            dst[i] = rhs[i];
+        }
+
+      // Ensure the distributed vector has the correct values everywhere
+      dst.compress(VectorOperation::insert);
+
+      rhs.resize(0); // remove rhs again
+    }
+}
+
+
+template <typename VectorType>
 void
-SparseDirectMUMPS::Tvmult(Vector<double> &dst, const Vector<double> &src) const
+SparseDirectMUMPS::Tvmult(VectorType &dst, const VectorType &src) const
 {
   // The matrix has at least one nonzero element:
   Assert(nnz != 0, ExcNotInitialized());
-
   Assert(n == dst.size(), ExcMessage("Destination vector has the wrong size."));
   Assert(n == src.size(), ExcMessage("Source vector has the wrong size."));
 
   id.icntl[8] = 2; // transpose
-  // Hand over right-hand side
-  copy_rhs_to_mumps(src);
-
-  // Start solver
-  id.job = 3;
-  dmumps_c(&id);
-  copy_solution(dst);
+  vmult(dst, src);
```

Comment:
This is quite difficult to understand. I *think* what is happening is that you set `id.ictnl[8]` to indicate the transpose, and then you run through the non-transpose function. This strikes me as wrong, because `vmult()` does not set `id.ictnl[8]`, and so you get wrong results if you did
```
  X.Tvmult (a,b);
  X.vmult(c,d);
```
The second call here will *also* execute a *transpose* vmult. I think that `vmult()` should correctly set `ictntl[8]` for its own operations, and `Tvmult()` should not call `vmult()`. If that leads to too much duplicated code, you can also put the common code into a shared helper function.